### PR TITLE
Y24-153 - Added volume check to PacBio pools

### LINKED ISSUE #1348: As a user (James W) I would like there to be a volume check for PacBio pools.
**User story**
As a user (James W) I would like there to be a volume check for PacBio pools.

**Who are the primary contacts for this story**
James W, Ben T, Seena N

**Who is the nominated tester for UAT**
James W

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] Expose available_volume and used_volume in the PacBio pools resource
- [ ] Validate available_volume when editing a pool, the same as libraries. This should be feature flagged.
- [ ] Validate used_volume when adding a pool to a run so it doesn't violate the volume, the same as libraries. This should be a separate feature flag.
- [ ] Check existing prod data doesn't violate the volume check (prod dump to local or use rails console on server)

**Additional context**
A lot of this work should be similar to the PacBio Libraries behaviour.

### PR DESCRIPTION
**Pull request by SHIV5T3R**
Closes #1348 

#### Changes proposed in this pull request

- A feature flagged change that validates `available_volume` when updating PacBio pools
    - The validation of available volume when editing a PacBio pool is feature flagged here

#### Additional Context

This PR leaves out adding attributes to the pool resource. When working on https://github.com/sanger/traction-ui/issues/1784, these will need to be added so that validation on the UI happens when adding a pool to a run (from what I can see, that validation already partially exists)

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

**Review by BenTopping — COMMENTED** [review]
Pretty much there just some minor comments.

Could we also add a test similar to https://github.com/sanger/traction-service/blob/develop/spec/models/pacbio/well_spec.rb#L138 but for checking pool volume.

File: lib/tasks/pacbio_data.rake [hunk 1]
```diff
@@ -54,8 +54,8 @@ namespace :pacbio_data do
           insert_size: library.insert_size, tag_id: library.tag_id, source_id: library.pacbio_request_id, source_type: 'Pacbio::Request', aliquot_type: :derived
         )
       end
-      Pacbio::Pool.create!(tube: Tube.create, used_aliquots:, volume: 1, concentration: 1, insert_size: 600, template_prep_kit_box_barcode: '029979102141700063023',
-                           primary_aliquot: Aliquot.new(volume: 1, concentration: 1, template_prep_kit_box_barcode: '029979102141700063023', insert_size: 500))
+      Pacbio::Pool.create!(tube: Tube.create, used_aliquots:, volume: 200, concentration: 1, insert_size: 600, template_prep_kit_box_barcode: '029979102141700063023',
+                           primary_aliquot: Aliquot.new(volume: 20, concentration: 1, template_prep_kit_box_barcode: '029979102141700063023', insert_size: 500))
```
Comment (BenTopping) [inline lib/tasks/pacbio_data.rake:58]: Primary aliquot volume and the pool volume should both be the same.

File: spec/factories/pacbio/pool.rb [hunk 2]
```diff
@@ -7,7 +7,7 @@
       library_factory { :pacbio_library }
     end
 
-    primary_aliquot { association :aliquot, source: instance, aliquot_type: :primary }
+    primary_aliquot { association :aliquot, source: instance, aliquot_type: :primary, volume: 10 }
```
Comment (BenTopping) [inline spec/factories/pacbio/pool.rb:10]: You can use the volume variable passed in so that the primary aliquot and pool have the same volume. You can update this to just `volume:` which is shorthand for `volume: volume`.

File: spec/factories/pacbio/pool.rb [hunk 3]
```diff
@@ -16,7 +16,7 @@
     used_aliquots do
       library_count.times.map do
         library = build(library_factory)
-        build(:aliquot, source: library, tag: library.tag, aliquot_type: :derived, used_by: instance)
+        build(:aliquot, source: library, tag: library.tag, aliquot_type: :derived, used_by: instance, volume: 10)
```
Comment (BenTopping) [inline spec/factories/pacbio/pool.rb:19]: Is this change needed?

File: spec/models/concerns/aliquotable_spec.rb [hunk 4]
```diff
@@ -121,6 +125,18 @@
       end
     end
 
+    context 'when primary aliquot volume for pools is less than used volume' do
+      it 'adds an error' do
+        skip unless Flipper.enabled?(:y24_153__enable_volume_check_pacbio_pool_on_update)
```
Comment (BenTopping) [inline spec/models/concerns/aliquotable_spec.rb:130]: I dont think this skip is needed anymore as you have set the flag to be always on at the top of the file

File: spec/models/pacbio/pool_spec.rb [hunk 5]
```diff
@@ -312,4 +313,14 @@
       expect(pool.sequencing_runs).to eq([plate1.run, plate2.run])
     end
   end
+
+  context 'before_update' do
+    it 'calls primary_aliquot_volume_sufficient method' do
+      skip unless Flipper.enabled?(:y24_153__enable_volume_check_pacbio_pool_on_update)
```
Comment (BenTopping) [inline spec/models/pacbio/pool_spec.rb:319]: Same comment as above.

File: spec/models/pacbio/well_spec.rb [hunk 6]
```diff
@@ -153,6 +157,30 @@
       well = build(:pacbio_well, used_aliquots: [build(:aliquot, source: library, volume: 100, aliquot_type: :derived)])
       expect(well).to be_valid
     end
+
+    it 'is not valid when using an invalid amount of volume from a pool' do
+      unless Flipper.enabled?(:y24_153__enable_volume_check_pacbio_pool_on_update)
```
Comment (BenTopping) [inline spec/models/pacbio/well_spec.rb:162]: same comment as above.

**Review by BenTopping — CHANGES_REQUESTED** [review]
Couple clean up comments but looks good.

File: app/models/pacbio/pool.rb [hunk 7]
```diff
@@ -24,16 +24,17 @@ class Pool < ApplicationRecord
...
     validate :used_aliquots_volume
+    before_update :primary_aliquot_volume_sufficient,
+                  if: -> { Flipper.enabled?(:y24_153__enable_volume_check_pacbio_pool_on_update) }
 
     accepts_nested_attributes_for :primary_aliquot
 
     def used_aliquots_volume
-      # Get all the aliquots that are libraries and have insufficient volume
+      # Get all the aliquots that are libraries or pools and have insufficient volume
       failed_aliquots = used_aliquots.select do |aliquot|
```
Comment (stevieing) [inline app/models/pacbio/pool.rb:34]: This method is exactly the same in pool and well. Maybe put it in aliquotable?

**Review by stevieing — CHANGES_REQUESTED** [review]

File: app/resources/v1/pacbio/pool_resource.rb [hunk 8]
```diff
@@ -75,6 +76,12 @@ def updated_at
       def publish_messages
         Messages.publish(@model.sequencing_runs, Pipelines.pacbio.message)
       end
+
+      def volume_tracking_attributes
+        return unless Flipper.enabled?(:y24_153__expose_volume_tracking_attributes_on_pool_resource)
+
+        attributes :used_volume, :available_volume
```
Comment (BenTopping) [inline app/resources/v1/pacbio/pool_resource.rb:83]: Testing locally, I dont think this is working. Seems to error when I try and access the pools endpoint with the flag on. I think we should be fine to remove this use of the feature flag entirely and just keep used_volume and available_volume exposed as the UI can now handle it.

**Review by BenTopping — APPROVED** [review]
Looks good.
We should hold off on merging until the current release is built (tomorrow afternoon).
